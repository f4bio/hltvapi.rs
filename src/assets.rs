use actix_web::{get, web, HttpResponse};
use include_dir::Dir;
use lazy_static::lazy_static;
use tera::Tera;
use tracing::{debug, error};

use crate::errors::ApiError;

pub static RESOURCES: Dir<'static> = include_dir!("./web/dist/");

lazy_static! {
  pub static ref TEMPLATES: Tera = {
    let mut tera = Tera::default();
    // let template_names = vec![
    //   "base.tera.html",
    //   "calendar.tera.html",
    //   "docs.tera.html",
    //   "landing.tera.html",
    //   "news.tera.html",
    //   "snackbar.tera.html",
    // ];
    tera.add_raw_templates(vec![
      (
        "base.tera.html",
        RESOURCES
          .get_file("base.tera.html")
          .unwrap()
          .contents_utf8()
          .unwrap(),
      ),
      (
        "calendar.tera.html",
        RESOURCES
          .get_file("calendar.tera.html")
          .unwrap()
          .contents_utf8()
          .unwrap(),
      ),
      (
        "docs.tera.html",
        RESOURCES
          .get_file("docs.tera.html")
          .unwrap()
          .contents_utf8()
          .unwrap(),
      ),
      (
        "landing.tera.html",
        RESOURCES
          .get_file("landing.tera.html")
          .unwrap()
          .contents_utf8()
          .unwrap(),
      ),
      (
        "news.tera.html",
        RESOURCES
          .get_file("news.tera.html")
          .unwrap()
          .contents_utf8()
          .unwrap(),
      ),
      (
        "snackbar.tera.html",
        RESOURCES
          .get_file("snackbar.tera.html")
          .unwrap()
          .contents_utf8()
          .unwrap(),
      ),
    ])
    .unwrap();
  tera.autoescape_on(vec!["html"]);
  // tera.register_filter("do_nothing", do_nothing_filter);
  tera
};
}

/// resource handler
#[get("/{resource}")]
pub async fn resources(resource: web::Path<String>) -> Result<HttpResponse, ApiError> {
  let response = if !resource.clone().is_ascii() {
    HttpResponse::Forbidden().body(format!(
      "invalid file name: '{}'",
      resource.clone().as_str()
    ))
  } else if !RESOURCES.contains(resource.clone().as_str()) {
    error!(
      "requested resource '{}' not found!",
      resource.clone().as_str()
    );
    HttpResponse::NotFound().body(format!("no such file: '{}'", resource.clone().as_str()))
  } else {
    // maybe unsafe?
    // -> template (*.tera.html) could be requested directly
    // TODO: double check this
    let resource_file = RESOURCES.get_file(resource.clone().as_str());
    let file_mime_guess = mime_guess::from_path(resource.clone().as_str());

    debug!(
      "file: {} mime: {}",
      resource_file.clone().unwrap().path,
      file_mime_guess.first().unwrap()
    );

    HttpResponse::Ok()
      .content_type(file_mime_guess.first().unwrap())
      .body(resource_file.unwrap().contents())
  };

  Ok(response)
}
