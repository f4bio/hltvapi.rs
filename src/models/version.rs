use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct Version {
  pub version: String,
}
