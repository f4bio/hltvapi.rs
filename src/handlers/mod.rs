use actix_web::{get, HttpRequest, HttpResponse};
use itconfig::get_env_or_default;
use rand::seq::SliceRandom;
use serde::Deserialize;
use tera::Context;

use crate::assets::TEMPLATES;
use crate::errors::ApiError;
use crate::models::version::Version;

pub mod api;

/// Limiter.top_tier: default (and currently only option): true
#[derive(Debug, Deserialize)]
pub struct Limiter {
  limit: Option<i64>,
  top_tier: Option<bool>,
}

impl Default for Limiter {
  fn default() -> Limiter {
    Limiter {
      limit: Some(10),
      top_tier: Some(true),
    }
  }
}

/// version handler
#[get("/version")]
pub async fn version() -> Result<HttpResponse, ApiError> {
  let app_version: String = String::from(env!("CARGO_PKG_VERSION"));

  Ok(
    HttpResponse::Ok()
      .content_type(mime::APPLICATION_JAVASCRIPT_UTF_8)
      .json(Version {
        version: app_version,
      }),
  )
}

/// calendar handler
#[get("/calendar")]
pub async fn calendar() -> Result<HttpResponse, ApiError> {
  let context = Context::new();
  let resp_body = TEMPLATES.render("calendar.tera.html", &context);

  Ok(
    HttpResponse::Ok()
      .content_type(mime::TEXT_HTML_UTF_8)
      .body(resp_body.unwrap()),
  )
}

/// docs handler
#[get("/docs")]
pub async fn docs() -> Result<HttpResponse, ApiError> {
  let context = Context::new();
  let resp_body = TEMPLATES.render("docs.tera.html", &context);

  Ok(
    HttpResponse::Ok()
      .content_type(mime::TEXT_HTML_UTF_8)
      .body(resp_body.unwrap()),
  )
}

/// news handler
#[get("/news")]
pub async fn news() -> Result<HttpResponse, ApiError> {
  let context = Context::new();
  let resp_body = TEMPLATES.render("news.tera.html", &context);

  Ok(
    HttpResponse::Ok()
      .content_type(mime::TEXT_HTML_UTF_8)
      .body(resp_body.unwrap()),
  )
}

/// index handler
#[get("/")]
pub async fn index(_req: HttpRequest) -> Result<HttpResponse, ApiError> {
  let wisdom_sentences = vec![
    "might get DMCA struck at all times...",
    "get 'em while they're hot...",
    "may go down every second...",
  ];
  let wisdom = wisdom_sentences.choose(&mut rand::thread_rng());
  let app_base_url: String = get_env_or_default("PUBLIC_BASE_URL", "https://hltvapi.f4b.io");

  let mut context = tera::Context::new();
  context.insert(
    "wisdom",
    wisdom.unwrap_or(&"might get DCMA struck at all times..."),
  );
  context.insert("app_base_url", app_base_url.as_str());
  let resp_body = TEMPLATES.render("landing.tera.html", &context);

  Ok(
    HttpResponse::Ok()
      .content_type(mime::TEXT_HTML_UTF_8)
      .body(resp_body.unwrap()),
  )
}

#[cfg(test)]
mod tests {
  use actix_web::http::StatusCode;
  use actix_web::{test, web, App, HttpResponse};

  // use super::index;

  #[actix_rt::test]
  async fn test_index_ok() {
    let app = test::init_service(
      App::new().service(web::resource("/").to(|| async { HttpResponse::Ok() })),
    )
    .await;

    // Create request object
    let req = test::TestRequest::with_uri("/").to_request();

    // Execute application
    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::OK);
    // insta::assert_yaml_snapshot!(resp.status().as_str());
  }
}
