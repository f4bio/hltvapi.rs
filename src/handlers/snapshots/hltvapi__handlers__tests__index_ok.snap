---
source: src/handlers/mod.rs
expression: resp.status().as_str()

---
"200"
